File: billing_library/src/main/java/com/robj/billinglibrary/BillingManager.java
package com.robj.billinglibrary;

import android.app.Application;
import android.content.Context;
import android.text.TextUtils;
import android.util.Log;

import java.util.concurrent.TimeUnit;

import static com.robj.billinglibrary.PrefsUtils.clearPref;
import static com.robj.billinglibrary.PrefsUtils.readLongPref;
import static com.robj.billinglibrary.PrefsUtils.readStringPref;
import static com.robj.billinglibrary.PrefsUtils.writeLongPref;
import static com.robj.billinglibrary.PrefsUtils.writeStringPref;

/**
 * Created by Rob J on 05/11/17.
 */

public class BillingManager {

    private static final String TAG = BillingManager.class.getSimpleName();

    private static final String TRIAL_START_DATE = "TRIAL_START_DATE";
    private static final String TRIAL_LENGTH = "TRIAL_LENGTH";
    private static final String PURCHASED_SKU = "PURCHASED_SKU";
    private static final String LAST_PURCHASED_CHECK_DATE = "LAST_PURCHASED_CHECK_DATE";

    protected BillingManager(Application context) {
        Billing.init(context);
    }

    private static long getTrialLengthInMillis(Context context) {
        return readLongPref(context, TRIAL_LENGTH);
    }

    protected void setTrialLengthInDays(Context context, int trialLengthInDays) {
        writeLongPref(context, TRIAL_LENGTH, TimeUnit.DAYS.toMillis(trialLengthInDays));
    }

    public static boolean isTrialStarted(Context context) {
        return getTrialLengthInMillis(context) > 0 && getTrialStartDate(context) > 0;
    }

    public static boolean isTrialPeriod(Context context) {
        if(isPurchased(context) || !isTrialStarted(context))
            return false;
        long diff = System.currentTimeMillis() - getTrialStartDate(context);
        return diff < getTrialLengthInMillis(context);
    }

    public static boolean hasPaidFeatures(Context context) {
        return isPurchased(context) || isTrialPeriod(context);
    }

    public static int getTrialPeriodLeft(Context context) {
        if(!isTrialPeriod(context))
            return 0;
        long diff = System.currentTimeMillis() - getTrialStartDate(context);
        diff = getTrialLengthInMillis(context) - diff;
        return (int) Math.ceil((double) diff/86400000);
    }

    public static long getTrialStartDate(Context context) {
        return readLongPref(context, TRIAL_START_DATE);
    }

    static void setTrialStartDate(Context context, long installDateInMillis) {
        writeLongPref(context, TRIAL_START_DATE, installDateInMillis);
    }

    public static boolean isPurchased(Context context) {
        String sku = readStringPref(context, PURCHASED_SKU);
        return !TextUtils.isEmpty(sku);
    }

    static void savePurchase(Context context, String sku) {
        if(sku == null) {
            clearPref(context, PURCHASED_SKU);
            clearPref(context, LAST_PURCHASED_CHECK_DATE);
        } else
            writeStringPref(context, PURCHASED_SKU, sku);
    }

    public static String getPurchasedSku(Context context) {
        return readStringPref(context, PURCHASED_SKU);
    }

    public static long getLastPurchasedCheckDate(Context context) {
        return readLongPref(context, LAST_PURCHASED_CHECK_DATE);
    }

    static void setLastPurchaseCheckedDate(Context context, long dateInMillis) {
        writeLongPref(context, LAST_PURCHASED_CHECK_DATE, dateInMillis);
    }

    public static void startTrial(Context context) {
        if(BillingManager.getTrialStartDate(context) == 0) {
            BillingManager.setTrialStartDate(context, System.currentTimeMillis());
            Log.d(TAG, "Starting trial..");
        } else
            Log.d(TAG, "Trial already started..");
    }

    public static void forceTrialExpiry(Context context) {
        setTrialStartDate(context, 1);
        Log.d(TAG, "Trial forcefully expired..");
    }

    public static class Builder {

        private final Application context;
        private int trialLengthInDays = 14;
        private boolean resetTrial = false;

        public Builder(Application context) {
            this.context = context;
        }

        public Builder setTrialLength(int trialLengthInDays) {
            if(trialLengthInDays <= 0)
                throw new RuntimeException("Trial length needs to be 1 day or more - default is 14");
            this.trialLengthInDays = trialLengthInDays;
            return this;
        }

        public Builder resetTrial(boolean resetTrial) {
            this.resetTrial = resetTrial;
            return this;
        }

        public BillingManager build() {
            BillingManager billingManager = new BillingManager(context);
            billingManager.setTrialLengthInDays(context, trialLengthInDays);
            if(resetTrial) {
                BillingManager.setTrialStartDate(context, System.currentTimeMillis());
                Log.d(TAG, "Trial reset..");
            }
            return billingManager;
        }

    }

}

File: billing_library/src/main/java/com/robj/billinglibrary/BillingManager.kt
package com.robj.billinglibrary

import android.app.Application
import android.content.Context
import android.text.TextUtils
import android.util.Log
import java.util.concurrent.TimeUnit

/**
 * Created by Rob J on 05/11/17.
 */

class BillingManager protected constructor(context: Application) {

    private val billing: Billing

    init {
        Billing.init(context)
        billing = Billing.getInstance(context) //Keep billing alive
    }

    protected fun setTrialLengthInDays(context: Context, trialLengthInDays: Int) {
        PrefsUtils.writeLongPref(context, TRIAL_LENGTH, TimeUnit.DAYS.toMillis(trialLengthInDays.toLong()))
    }

    class Builder(private val context: Application) {
        private var trialLengthInDays = 14
        private var resetTrial = false

        fun setTrialLength(trialLengthInDays: Int): Builder {
            if (trialLengthInDays <= 0)
                throw RuntimeException("Trial length needs to be 1 day or more - default is 14")
            this.trialLengthInDays = trialLengthInDays
            return this
        }

        fun resetTrial(resetTrial: Boolean): Builder {
            this.resetTrial = resetTrial
            return this
        }

        fun build(): BillingManager {
            val billingManager = BillingManager(context)
            billingManager.setTrialLengthInDays(context, trialLengthInDays)
            if (resetTrial) {
                BillingManager.setTrialStartDate(context, System.currentTimeMillis())
                Log.d(TAG, "Trial reset..")
            }
            return billingManager
        }

    }

    companion object {

        private val TAG = BillingManager::class.java.simpleName

        private val TRIAL_START_DATE = "TRIAL_START_DATE"
        private val TRIAL_LENGTH = "TRIAL_LENGTH"
        private val PURCHASED_SKU = "PURCHASED_SKU"
        private val LAST_PURCHASED_CHECK_DATE = "LAST_PURCHASED_CHECK_DATE"

        private fun getTrialLengthInMillis(context: Context): Long {
            return PrefsUtils.readLongPref(context, TRIAL_LENGTH)
        }

        @JvmStatic
        fun isTrialStarted(context: Context): Boolean {
            return getTrialLengthInMillis(context) > 0 && getTrialStartDate(context) > 0
        }

        @JvmStatic
        fun isTrialPeriod(context: Context): Boolean {
            if (isPurchased(context) || !isTrialStarted(context))
                return false
            val diff = System.currentTimeMillis() - getTrialStartDate(context)
            return diff < getTrialLengthInMillis(context)
        }

        @JvmStatic
        fun hasPaidFeatures(context: Context): Boolean {
            return isPurchased(context) || isTrialPeriod(context)
        }

        @JvmStatic
        fun getTrialPeriodLeft(context: Context): Int {
            if (!isTrialPeriod(context))
                return 0
            var diff = System.currentTimeMillis() - getTrialStartDate(context)
            diff = getTrialLengthInMillis(context) - diff
            return Math.ceil(diff.toDouble() / 86400000).toInt()
        }

        @JvmStatic
        fun getTrialStartDate(context: Context): Long {
            return PrefsUtils.readLongPref(context, TRIAL_START_DATE)
        }

        internal fun setTrialStartDate(context: Context, installDateInMillis: Long) {
            PrefsUtils.writeLongPref(context, TRIAL_START_DATE, installDateInMillis)
        }

        @JvmStatic
        fun isPurchased(context: Context): Boolean {
            val sku = PrefsUtils.readStringPref(context, PURCHASED_SKU)
            return !TextUtils.isEmpty(sku)
        }

        @JvmStatic
        fun savePurchase(context: Context, sku: String?) {
            if (sku == null) {
                PrefsUtils.clearPref(context, PURCHASED_SKU)
                PrefsUtils.clearPref(context, LAST_PURCHASED_CHECK_DATE)
            } else
                PrefsUtils.writeStringPref(context, PURCHASED_SKU, sku)
        }

        @JvmStatic
        fun getPurchasedSku(context: Context): String {
            return PrefsUtils.readStringPref(context, PURCHASED_SKU)
        }

        @JvmStatic
        fun getLastPurchasedCheckDate(context: Context): Long {
            return PrefsUtils.readLongPref(context, LAST_PURCHASED_CHECK_DATE)
        }

        internal fun setLastPurchaseCheckedDate(context: Context, dateInMillis: Long) {
            PrefsUtils.writeLongPref(context, LAST_PURCHASED_CHECK_DATE, dateInMillis)
        }

        @JvmStatic
        fun startTrial(context: Context) {
            if (BillingManager.getTrialStartDate(context) == 0L) {
                BillingManager.setTrialStartDate(context, System.currentTimeMillis())
                Log.d(TAG, "Starting trial..")
            } else
                Log.d(TAG, "Trial already started..")
        }

        @JvmStatic
        fun forceTrialExpiry(context: Context) {
            setTrialStartDate(context, 1)
            Log.d(TAG, "Trial forcefully expired..")
        }
    }

}
